client/go/outline/iptable/packet_listener.go

```diff
@@ -0,0 +1,319 @@
+	}, nil
+}
+
+// SetDefault sets the [transport.PacketListener] to be used for IP addresses that
+// do not match any specific rule in the IP table.
+// Unlike the constructor, this method allows `defaultListener` to be nil, which
+// would mean no packets can be sent or received for addresses not in the table.
+func (listener *PacketListener) SetDefault(defaultListener transport.PacketListener) {
+	listener.defaultListener = defaultListener
+}
+
+// ListenPacket returns a [net.PacketConn] that routes packets based on IP addresses.
+// The returned connection aggregates packets from and dispatches packets to various
+// underlying listeners as defined by the IP table and the default listener.
+func (listener *PacketListener) ListenPacket(ctx context.Context) (net.PacketConn, error) {
+	return newPacketConn(ctx, listener.table, listener.defaultListener)
+}
+
+const packetQueueSize = 128
+
+type incomingPacket struct {
+	data []byte
+	addr net.Addr
+}
+
+// packetConn is a [net.PacketConn] implementation that multiplexes
+// packets over various underlying [net.PacketConn]s based on IP addresses.
+// It is returned by [PacketListener.ListenPacket].
+//
+// For outgoing packets (WriteTo), it determines the appropriate underlying
+// connection using an [IPTable] and a default listener. New underlying
+// connections are established on-demand.
+// For incoming packets (ReadFrom), it aggregates packets received from all
+// active underlying connections (both specific and default) into a single channel.
+type packetConn struct {
+	defaultListener transport.PacketListener
+	defaultConn     net.PacketConn
+
+	listenerTable IPTable[transport.PacketListener]
+	connMap       map[netip.Addr]net.PacketConn
+	connMapLock   sync.Mutex
+
+	forwardingContext            context.Context
+	closePacketForwardingContext context.CancelFunc
+	forwardedPackets             chan incomingPacket
+	forwardCounter               sync.WaitGroup
+
+	lastRemainingPacket     incomingPacket
+	lastRemainingPacketLock sync.Mutex
+}
+
+var _ net.PacketConn = (*packetConn)(nil)
+
+func newPacketConn(
+	parentContext context.Context,
+	listenerTable IPTable[transport.PacketListener],
+	defaultListener transport.PacketListener,
+) (*packetConn, error) {
+	forwardingContext, closeForwardingContext := context.WithCancel(parentContext)
+
+	var defaultConn net.PacketConn
+	if defaultListener != nil {
+		connAttempt, err := defaultListener.ListenPacket(forwardingContext)
+
+		if err != nil {
+			defaultConn = nil
+		} else {
+			defaultConn = connAttempt
+		}
+	}
+
+	conn := &packetConn{
+		closePacketForwardingContext: closeForwardingContext,
+		connMap:                      make(map[netip.Addr]net.PacketConn),
+		connMapLock:                  sync.Mutex{},
+		defaultConn:                  defaultConn,
+		defaultListener:              defaultListener,
+		forwardedPackets:             make(chan incomingPacket, packetQueueSize),
+		forwardingContext:            forwardingContext,
```

Comment:
*nit* Can we put related fields together? For example, `forwardingContext` can be put with `closePacketForwardingContext`.